{% extends "base.html" %}
{% block content %}

<!--    <div class="contents local topic" id="contents">
        <ul class="simple">
            <li> 1 - <a class="reference internal" href="#Stats" id="id1">
                Stats</a></li>
            <li> 2 - <a class="reference internal" href="#Dependencies"
                id="id11">Dependencies</a> </li>
        </ul>
    </div>
-->
<p>
Sequencing depth and genome coverage report <br>
<b>Chromosome:</b> {{chrom_name}} 
{% if project %}
<b>Project :</b> {{project_name}}
{% endif %}
</p>

<!--This page informs of the mapping depth coverage. Some regions are detected
    with an unusual depth coverage and are showed below.-->

<div class="section" id="Coverage">
    <h2><a class="toc-backref" href="#id1">Coverage</a></h2>
    <div>
    The following figure shows the per-base coverage along the reference genome
(black line). The blue line indicates the running median. From the normalised
coverage, we estimate z-scores on a per-base level. The red lines indicates the
z-scores at plus or minus N standard deviations, where N is chosen by the user
(default:4).

    </div>
    <img src={{ cov_plot }}> 
</div>

<div class="section" id="Submapping list"> 
The image above is a static image. However, interactive plots are available from
the table here below. To allows interactions, we split the data into  chunks of 
500,000 bp length at max. The names contains the start and end positions.
    {{ list_submapping }}
</div>

<div class="section" id="histogram">    

<div>The following figure contains the histogram of the genome coverage. The X and Y
axis being in log scale.</div>
<img style="width:45%"  src={{ cov_hist_loglog }}> 
<img style="width:45%" src={{ cov_hist_logy }}> 
</div>
    



<div class="section" id="Basic stats">
    <h2><a class="toc-backref" href="#id1">Basic stats</a></h2>
        <p>The following table gives some basic statistics about the genome coverage. 
            {{ nc_stats }}
</div>



<h2><a class="toc-backref" href="#id1">Regions of Interest</a></h2>
Running median is the median computed along the genome using a sliding window.
The following tables give regions of interest detected by sequana. Here is some captions:
<ul>
    <li>mean_cov: the average of coverage</li>
    <li>mean_rm: the average of running median</li>
    <li>mean_zscore: the average of zscore</li>
    <li>max_zscore: the higher zscore contains in the region</li>
</ul>

<div class="section" id="Low coverage region">
    <h3><a class="toc-backref" href="#id1">Low coverage region</a></h3>
    {{ lc_paragraph }}
    {{ low_coverage }}
</div>

<div class="section" id="High coverage region">
    <h3><a class="toc-backref" href="#id1">High coverage region</a></h3>
    {{ hc_paragraph }}
    {{ high_coverage }}
</div>


{% if coverage_vs_gc %}
<div class="section">
    <h2>Coverage versus GC</h2>
    {{ coverage_vs_gc }}
</div>
{% endif %}

<div class="section" id="Normalized coverage">
    <h2><a class="toc-backref" href="#id1">Normalized coverage</a></h2>
    {{ nc_paragraph }}
    <div>
        <img src={{ nc_plot }}> </>
    </div>
</div>

<div class="section" id="Z-Score distribution">
    <h2><a class="toc-backref" href="#id1">Z-Score distribution</a></h2>
    {{ bp_paragraph }}
    <div>    
        <img src={{ bp_plot }}> </>
    </div>
</div>



{% endblock %}
